File: app/frontend/src/api/game.js
const BASE_URL_WS = import.meta.env.VITE_API_WS;
const BASE_URL_API = import.meta.env.VITE_API_URL;
// Paths
const gameMovePath = `${BASE_URL_WS}/game/client-socket`;
const gameSearchPath = `${BASE_URL_API}/game/match-search`;
const gameCancelPath = `${BASE_URL_API}/game/match-search/cancel`;


var ws = new WebSocket(gameMovePath);

export function sendPieceMove(moveStr) {
    try {
        const data = {
            "move_str": moveStr
        }
        ws.send(JSON.stringify(data));

        return new Promise((resolve, reject) => {
            if (ws.readyState !== WebSocket.OPEN) {
                return reject("WebSocket not open");
            }

            ws.addEventListener("message", (event) => {
                const data = JSON.parse(event.data);
                const newBoardState = data.boardState;
                const gameState = data.gameState;
                resolve({ boardState: newBoardState, gameState: gameState});
            }, { once: true });
        });
    } catch (err) {
        console.log("[ERROR] game.js/sendPieceMove: " + err);
    }
}


export async function searchForGame(userID) {
    const response = await fetch(gameSearchPath,
        {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                "client_uid": userID,
            })
        }
    );
    const data = await response.json();
    return data;
}


export async function cancelSearch(userID) {
    const response = await fetch(gameCancelPath,
        {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                "client_uid": userID,
            })
        }
    );
    const data = await response.json();
    return data;
}

File: app/backend/api/endpoints/game.py
import json
from fastapi import WebSocket, APIRouter
from services.chess_logic import move_auth
from services.match_pool import MatchPool
from api.schemas import Matchmaking


game_router = APIRouter()


mp = MatchPool()

@game_router.post("/game/match-search")
async def match_search(data: Matchmaking):
    uid: str = data.client_uid

    match_state: dict = mp.add_plyr(uid)
    return match_state


@game_router.post("/game/match-search/cancel")
async def cancel_search(data: Matchmaking):
    uid: str = data.client_uid

    request_outcome: dict = mp.cancel_search(uid)
    return request_outcome


@game_router.websocket("/game/client-socket")
async def game_socket(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        data = json.loads(data)
        
        move_str: str = data["move_str"]

        new_board_state, game_state = move_auth(move_str)
        await websocket.send_text(json.dumps({
            "boardState": new_board_state,
            "gameState": game_state}))